libafl/src/mutators/mutations.rs

```diff
@@ -1200,6 +1194,153 @@ impl<I> CrossoverReplaceMutator<I> {
         }
     }
 }
+/// Crossover insert mutation for inputs mapped to a bytes vector
+#[derive(Debug)]
+pub struct MappedCrossoverInsertMutator<'a, F> {
+    input_from_corpus_mapper: &'a F,
+}
+
+impl<'a, S, F, I> Mutator<I, S> for MappedCrossoverInsertMutator<'a, F>
+where
+    S: HasCorpus + HasMaxSize + HasRand + UsesInput,
+    I: HasMutatorBytes,
+    F: for<'b> Fn(&'b S::Input) -> Option<&'b Vec<u8>>,
```

Comment:
Can't you just implement `HasMutatorBytes` for `S::Input` and achieve the exact same thing that the Mapped mutator does?

Reply:
No, I want to be able to use these Mutators on different parts of custom inputs. Implementing `HasMutatorBytes` would mean that I'd have to statically decide which part is extracted for the mutator, while this implementation allows doing this dynamically. This is the entire reason for this PR.

Check out the example fuzzer, this may make clear what I'm trying to accomplish.

Reply:
You can create an input from a Vec<u8> using `MutVecInput`, and from there you can use `.sub_input()` to get a sub input for a certain range - dynamically

Reply:
I guess I can see a reason to randomly map a part of a complex input to a sub_input(?)

Reply:
This PR started based on a fuzzer I built to test programs with a command line interface. I modelled all the input they take (such as `stdin`, command line flags without attributes, command line args with associated binary data, etc.) in a custom `Input` type consisting of several fields in the struct, one for each attribute/stdin/etc. The example fuzzer I added in this PR is based on this.

Now, this complex `Input` contained multiple byte arrays (i.e. `Vec<u8>`), and I wanted to use `havoc_mutations` on each part. This is the easiest implementation to get this to work, without having to duplicate all the mutators within `havoc_mutations`.

Check out #2202 for an extensive discussion about this, and look at the example fuzzer for the interface I want to provide.

Reply:
If you're curious, this is the fuzzer in question: https://github.com/riesentoaster/coreutils-differential-fuzzing

This PR would significantly simplify the code additions required to test the other utils.

Reply:
Ok I get what you use the mapping for. I think it's good, but we shouldn't use `Vec<u8>` explicitly. For example, if we have a HasMutatorBytes you could have an input out of a single `Vec` with multiple mapper functions for it

Reply:
Thinking this further, HasMutatorBytes is basically one special case of a mapping...

Reply:
Also, @addisoncrump I think multipart mutators could then be moved over to use mappings instead(?)
That way we don't have two "ecosystems" for the same thing

Reply:
I'd prefer if the interface of `havoc_mutations_mapped` would be consistent, i.e. either have two mappers returning a `Vec<u8>` or a `MutVecInput`. Since it's usually easier to get a Vec<u8> (so users of `havoc_mutations_mapped` don't need to convert it themselves) and at least for the immutable mapper I'd need to extract a `Vec<u8>` reference either way, I went with that.

Also, getting a mutator for `Vec<u8>` based on one for `HasMutatorBytes` is trivial with the new system; that's what `MutVecMappingMutator` is for.

As an aside: The user currently needs to pass essentially the same function to `havoc_mutations_mapped` *twice*, once with mutable and once with immutable references. I didn't find a simpler way to pass Rust's compiler checks, but since it's essentially the same logic, this is a bit of a shame. You wouldn't know of a way to do that?

Reply:
Regarding multipart: It may be slightly more efficient, because if may have fewer function calls, but on the other hand requires an additional random number to choose between subparts and allows less control for the mutation scheduler. And of course if you have *many* parts, my system would introduce quite a bit of additional and very similar/duplicate-looking code for the extraction functions.

I haven't done any performance tests so far, but this may be interesting.

Reply:
The reason I'm pushing for `HasMutatorBytes` is that it works for sub parts of a Vec, or even completely non-vec-inputs, while Vec<u8> and even `MutVecInput` don't.
If you want I can try to fiddle this in(?)

Reply:
I mean, if you can implement `HasMutatorBytes` for your custom Input, you can just use the default `havoc_mutations`.

It may be possible to add an additional version of `havoc_mutations` where the passed functions don't extract `Vec<u8>` but anything implementing `HasMutatorBytes`. In the background, I'd probably use the existing version, and just extract the `Vec<u8>` based on `HasMutatorBytes`. So it'd essentially just be a wrapper around the current `havoc_mutations_mapped`.

Reply:
> I mean, if you can implement HasMutatorBytes for your custom Input, you can just use the default havoc_mutations.

Same argument as you: You may not want to mutate the full input, but only parts of it, hence the mapper makes sense here as well.

> In the background, I'd probably use the existing version, and just extract the Vec<u8> based on HasMutatorBytes

The issue is that turning `HasMutatorBytes` into a Vec needs an allocation, turning Vec into `HasMutatorBytes` is free.

Reply:
Ah wait I got an idea, we can turn the logic around. Let me try...